File: src/Calculator/Utils/DispatcherTimerDelayer.cpp
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

#include "pch.h"
#include "DispatcherTimerDelayer.h"

using namespace CalculatorApp;
using namespace Windows::UI::Xaml;
using namespace Windows::Foundation;

DispatcherTimerDelayer::DispatcherTimerDelayer(TimeSpan timeSpan)
{
    m_timer = ref new DispatcherTimer();
    m_timer->Interval = timeSpan;
    auto interval = m_timer->Interval;
    m_timer->Tick += ref new EventHandler<Object ^>(this, &DispatcherTimerDelayer::Timer_Tick);
}

void DispatcherTimerDelayer::Timer_Tick(Object ^ sender, Object ^ e)
{
    m_timer->Stop();
    Action(this, nullptr);
}

void DispatcherTimerDelayer::Start()
{
    m_timer->Start();
}

void DispatcherTimerDelayer::ResetAndStart()
{
    m_timer->Stop();
    m_timer->Start();
}

void DispatcherTimerDelayer::Stop()
{
    m_timer->Stop();
}

File: src/Calculator/Utils/DispatcherTimerDelayer.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System;

using Windows.UI.Xaml;

namespace CalculatorApp
{
    public sealed class DispatcherTimerDelayer
    {
        public event EventHandler<object> Action;

        public DispatcherTimerDelayer(TimeSpan timeSpan)
        {
            m_timer = new DispatcherTimer
            {
                Interval = timeSpan
            };
            var interval = m_timer.Interval;
            m_timer.Tick += Timer_Tick;
        }

        public void Start()
        {
            m_timer.Start();
        }

        public void ResetAndStart()
        {
            m_timer.Stop();
            m_timer.Start();
        }

        public void Stop()
        {
            m_timer.Stop();
        }

        private void Timer_Tick(object sender, object e)
        {
            m_timer.Stop();
            Action?.Invoke(this, null);
        }

        private readonly DispatcherTimer m_timer;
    }
}
